### PR TITLE
add rcu lock, and use it for cluster update

### PR DESCRIPTION
**Pull request by nejisama**
### Issues associated with this PR

#345 

### Sign the CLA

done

### Solutions

1. implement an rcu lock 
2. primaryCluster adds an rcu lock for configUsed,  ClusterManager GetSnapshot will "lock" it
3. When updates the cluster, the cluster will be updated directly, but the update api will update the config's rcu lock, and returns until the rcu lock update success/timeout
4. The cluster snapshot will put back to clustermanger to release the "config lock", makes the config can update directly.
5. use router chain

### UT result

pass

### Benchmark

+ Total TPS 14,000 
+ 5 machines
+ Add a test http server to call UpdateHosts(currently, mosn admin api does not support do this)
+ Update success return: 78ms~1.6s

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/proxy/downstream.go [hunk 1]
```diff
@@ -749,7 +760,7 @@ func (s *downStream) setupRetry(endStream bool) bool {
 
 // Note: retry-timer MUST be stopped before active stream got recycled, otherwise resetting stream's properties will cause panic here
 func (s *downStream) doRetry() {
-	pool, err := s.initializeUpstreamConnectionPool(s.cluster.Name(), nil)
+	pool, err := s.initializeUpstreamConnectionPool(s)
```
Comment (taoyuanyuan) [inline pkg/proxy/downstream.go]: Should putClusterSnapshot and Get again？

Comment (nejisama) [reply pkg/proxy/downstream.go]: Get is called in doReceiveHeaders

File: pkg/upstream/cluster/clustermanager.go [hunk 2]
```diff
@@ -191,23 +245,28 @@ func (cm *clusterManager) loadCluster(clusterConfig v2.Cluster, addedViaAPI bool
...
+	cm.primaryClusters.Store(clusterConfig.Name, NewPrimaryCluster(cluster, &clusterConfig, addedViaAPI))
 
 	return true
 }
 
-func (cm *clusterManager) getOrCreateClusterSnapshot(clusterName string) *clusterSnapshot {
+func (cm *clusterManager) PutClusterSnapshot(snapshot types.ClusterSnapshot) {
+	if s, ok := snapshot.(*clusterSnapshot); ok {
+		s.value.Put(s.config)
+	}
```
Comment (ldxdl) [inline pkg/upstream/cluster/clustermanager.go]: 这里加一个else err信息打印

File: pkg/upstream/cluster/clustermanager.go [hunk 3]
```diff
@@ -124,6 +127,60 @@ type primaryCluster struct {
...
+		addedViaAPI: addedViaAPI,
+		configUsed:  config,
+		updateLock:  sync.Mutex{},
+		configLock:  rcu.NewValue(config),
+	}
+}
+func (pc *primaryCluster) UpdateCluster(cluster types.Cluster, config *v2.Cluster, addedViaAPI bool) error {
+	pc.updateLock.Lock()
+	defer pc.updateLock.Unlock()
+	if cluster != nil {
```
Comment (ldxdl) [inline pkg/upstream/cluster/clustermanager.go]: 这里没看懂为啥要判断cluster，config为nil；是否允许为nil，是否允许同时为nil，这么写很容易静默错误

File: pkg/upstream/cluster/clustermanager.go [hunk 4]
```diff
@@ -124,6 +127,60 @@ type primaryCluster struct {
...
+	config.Hosts = hostConfig
+	pc.configUsed = config
+	if err := pc.configLock.Update(pc.configUsed, 0); err == rcu.Block {
+		return err
+	}
+	return nil
+}
+
+func deepCopyCluster(cluster *v2.Cluster) *v2.Cluster {
+	if cluster == nil {
```
Comment (ldxdl) [inline pkg/upstream/cluster/clustermanager.go]: 内部方式是否需要防御？建议有统一的防御思路，我现在看下来接口入口没有防御，但是内部会有一些防御的代码，这个思路需要再统一一下

File: pkg/types/upstream.go [hunk 5]
```diff
@@ -38,21 +38,21 @@ type ClusterManager interface {
 
 	SetInitializedCb(cb func())
 
-	// Clusters, return all cluster belongs to this clustermng
-	Clusters() map[string]Cluster
-
 	// Get, use to get the snapshot of a cluster
 	Get(context context.Context, cluster string) ClusterSnapshot
```
Comment (ldxdl) [inline pkg/types/upstream.go]: 这里是不是跟put对应，把名字也改成GetClusterSnapshot比较合适？

File: pkg/router/types.go [hunk 6]
```diff
@@ -198,4 +197,4 @@ func (p *routerPolicy) LoadBalancerPolicy() types.LoadBalancerPolicy {
 	return nil
 }
 
-type RouterFactory func(header []v2.HeaderMatcher) RouteBase
+type RouterRuleFactory func(base *RouteRuleImplBase, header []v2.HeaderMatcher) RouteBase
```
[the remark this answers is not in the document]
Comment (nejisama) [reply pkg/router/types.go]: Factory类的接口，都没有放在types里

File: pkg/upstream/cluster/clustermanager.go (hunk 4 above)
Comment (nejisama) [reply pkg/upstream/cluster/clustermanager.go]: 这个是一个工具类的方法 是期望nil就返回nil，非nil返回一个拷贝，nil也应该是一个合法的输入，只是处理不一样

**Comment by ldxdl** [comment]
LGTM

**Pull request merged**